File: solutions/169_majority_element.rb

# Given an array of size n, find the majority element. The majority element is the element that appears more than ⌊ n/2 ⌋ times.
#
# You may assume that the array is non-empty and the majority element always exist in the array.
#
# Credits:
# Special thanks to @ts for adding this problem and creating all test cases.

# @param {Integer[]} nums
# @return {Integer}
def majority_element(nums)
  nums.sort[nums.size / 2]
end

def majority_element2(nums)
  h = {}
  nums.each do |n|
    if h.key?(n)
      return n if h[n] + 1 > nums.size / 2.0
      h[n] += 1
    else
      h[n] = 1
    end
  end
  nums[0]
end

File: python_solutions/169.majority-element.py
from collections import Counter, defaultdict, OrderedDict, deque
from bisect import bisect_left, bisect_right 
from functools import reduce, lru_cache 
from typing import List 
import itertools 
import math 
import heapq 
import string
true = True
false = False
MIN, MAX = -0x3f3f3f3f, 0x3f3f3f3f
#
# @lc app=leetcode id=169 lang=python3
#
# [169] Majority Element
#
# https://leetcode.com/problems/majority-element/description/
#
# algorithms
# Easy (56.53%)
# Total Accepted:    546.6K
# Total Submissions: 966.5K
# Testcase Example:  '[3,2,3]'
#
# Given an array of size n, find the majority element. The majority element is
# the element that appears more than ⌊ n/2 ⌋ times.
# 
# You may assume that the array is non-empty and the majority element always
# exist in the array.
# 
# Example 1:
# 
# 
# Input: [3,2,3]
# Output: 3
# 
# Example 2:
# 
# 
# Input: [2,2,1,1,1,2,2]
# Output: 2
# 
# 
#
class Solution:
    def majorityElement(self, nums: List[int]) -> int:
        nums.sort()
        return nums[len(nums) // 2]

sol = Solution()
ns = [3,2,3]
ns = [8,9,8,9,8]
print(sol.majorityElement(ns))
